playback/src/audio_backend/alsa.rs

```diff
@@ -107,12 +116,9 @@ impl Sink for AlsaSink {
         if self.pcm.is_none() {
             let pcm = open_device(&self.device, self.format);
             match pcm {
-                Ok((p, period_size)) => {
+                Ok((p, buffer_size_bytes)) => {
                     self.pcm = Some(p);
-                    // Create a buffer for all samples for a full period
-                    self.buffer = Vec::with_capacity(
-                        period_size as usize * BUFFERED_PERIODS as usize * self.format.size(),
-                    );
+                    self.buffer = Vec::with_capacity(buffer_size_bytes);
```

Comment:
Err... OK, that old code clearly doesn't match its comment. I don't remember this going in but it makes no sense to me. Have you gone through the history of this file and seen all the changes, and more importantly, why they were made so we might understand the context of some of this weirdness?

Reply:
I did I. It was introduced in https://github.com/librespot-org/librespot/commit/770ea15498a0f1cfc7b9986f0954f4150258c29f but the more you dig the more you see that it was a mess to begin with IMHO. It has never been correct as far as I can tel,lol!!!

I think the original intention was to have `self.buffer` be only the size of a Period. I think @roderickvd's intention was to make the buffer the size of all Periods, so about 0.5 secs. To me making the buffer the total size of the buffer as apposed to just a period makes sense as it doesn't cost much and helps safe guard against underruns.

I'm not completely sure how the buffer works though? As I understand it `self.buffer` can grow to whatever size it needs to be. I'm not sure for example if ALSA blocks when it's buffer is full or what?

Reply:
It worked in the past not because it was correct but because ALSA is pretty forgiving as far as accepting a wide range of parameters and the internal buffer wasn't small enough to cause underruns as far as I can tell.

Reply:
>  if ALSA blocks when it's buffer is full or what

Yes, we are using it in blocking mode. That blocking is what keeps librespot's song position correct.

Reply:
> Yes, we are using it in blocking mode. That blocking is what keeps librespot's song position correct.

Then having a nice fat internal buffer surely wouldn't hurt anything then?

Reply:
To really understand the internal buffer I'd like to hear from @sniperrifle2004 as they wrote it. I'm really looking for a basic rational and the pros and cons of having the buffer be a period or the size of the whole buffer (ideally 4 periods). I notice no difference between the 2 except maybe ever so slightly less CPU usage on my Pi Zero when the buffer is the size of the entire buffer.

Reply:
Well I can tell you my rationale: I was noticing a lot of underruns on alsa-pulse with my raspberry pi. I had a bit of a look and noticed that aplay did not stutter at all. After a while I noticed that the main difference was that while librespot set a buffer size it did not set a period size. The original PR contains this same explanation:

https://github.com/librespot-org/librespot/pull/493


I looked into it more a bit later for a contribution to RustAudio/cpal and it turns out that by default the value for the period size is the minimum of the possible period size. The effect: Basically a low wakeup interval with high latency.

Now as to the buffer being a period: I don't think that matters as much, but it worked to write at least period sized chunks to alsa. Using full buffers should work as well, especially in blocking write mode. Just nothing smaller than a period size, since that can only have a negative effect.

Reply:
> I think the original intention was to have `self.buffer` be only the size of a Period. I think @roderickvd's intention was to make the buffer the size of all Periods, so about 0.5 secs. To me making the buffer the total size of the buffer as apposed to just a period makes sense as it doesn't cost much and helps safe guard against underruns.

Yes, this is what I did and never looked back because it worked well.
 
> I'm not completely sure how the buffer works though? As I understand it `self.buffer` can grow to whatever size it needs to be. I'm not sure for example if ALSA blocks when it's buffer is full or what?

Yes, `Vec::with_capacity(x)` means to initialize a `Vec` and allocate memory `x` elements in one swoop. It does not mean it is limited to `x` elements. Rather, it is an optimization to prevent frequent memory reallocations.

At the moment the `librespot` buffer is only written to Alsa when the `librespot` buffer is full:
https://github.com/librespot-org/librespot/blob/83e3eb8cbbf89f930af9080c7f58c1d374f02e3c/playback/src/audio_backend/alsa.rs#L162-L164

Suppose that would happen at the same speed as Alsa drains its own buffer, well then that'd be an interesting flip-flop between the Alsa buffer being completely full or just empty (and risking underrun). It's probably a good idea to make the `librespot` buffer smaller, with more frequent writes, to rule out that possibility.

> Now as to the internal buffer being a period: I don't think that matters as much, but it worked to write at least period sized chunks to alsa. Using full buffers should work as well, especially in blocking write mode. Just nothing smaller than a period size, since that can only have a negative effect.

Thanks for that PR and coming back to it after this time.

I think the optimal `librespot` buffer size can only be determined experimentally and when under load. Taking cue from `aplay` seems pretty reference to me.

Reply:
So the consensus seems to be period sized as it was before? I don't really have a preference. I'll look around though and see what others are doing.

Reply:
> NOTE: alsa-pulse advertises a ridiculously low minimum period size.

Using the alsa backend though PipeWire on Fedora 35 (x86_64) and `cat /proc/asound/card*/pcm*p/sub*/hw_params` does show a really small period size of 512 but a huge buffer of 174762 frames but throwing some `println!`'s in `alsa.rs` shows that it's giving us the 4 periods at 5512 frames pretty much just like we're asking for. `cat /proc/asound/card*/pcm*p/sub*/hw_params`  though PulseAudio on Mint 20 (x86_64) goes the opposite direction with a really big but gives us 4 periods at 5512. The discrepancy is probably due to both being software mixers. `cat /proc/asound/card*/pcm*p/sub*/hw_params` most likely is reporting PipeWire/PulseAudio to hardware buffer sizes and we're getting librespot > PipeWire/PulseAudio buffer sizes. Basically there's 2 buffers in that case.

So that's basically the 3 endpoints(?) that the alsa backend is likely to encounter, bare ALSA, PipeWire and PulseAudio and they all give us basically what we want. The only wildcard is someone manually configuring things in `/etc/asound.conf` which that's really on them.

Reply:
The more I think about it, the more I think my vote is to not reinvent the wheel. @sniperrifle2004's PR fixed the issue they were having and no one else has complained about have they? I say stick to what seems to work, a period size buffer.

Reply:
> I looked into it more a bit later for a contribution to RustAudio/cpal and it turns out that by default the value for the period size is the minimum period size. The effect: Basically a low wakeup interval with high latency.

The low wakeup interval makes sense. If you set a reasonable sized buffer with min period size that would make for ***a lot*** of periods. I'm not sure why it would effect latency though? The total buffer size is the latency, so a 0.5 sec buffer with 4 periods or 10 periods is still just 0.5 sec of latency. 

> Just nothing smaller than a period size, since that can only have a negative effect.

I haven't tested it on my Pi Zero but on my desktop it causes a weird distorted tremolo effect. I would not advise,lol!!!

Reply:
OK so to summarise what is going on in it's entirety:
* We are configuring ALSA with 4 periods of 5513 frames (approx 125ms) so there will be 4 hardware interupts per ALSA ring buffer.
* Internally we are also buffering 22052 frames (500ms) worth of data.
* We write our entire buffer (22052 frames) at a time to ALSA, rather than one period-worth (5513 frames) - arguably a bit odd.
* Once the hardware has played the first period of frames, it starts on the second period and there's an interupt to notify there's now space for the next period of data to be written
* But we then try to write four periods of data and `io.writei()` blocks until there is space, which in turn ripples back and blocks all of librespot's playback (which is fine).
* Only once everything has played out (during which time 3 more interupts were issued) is there space for `io.writei()` to finish and return.
* We then refill our internal buffer which we can presumably do way before the ring buffer empties. 

This does seem to work OK and I think maybe that's because we can re-fill our buffer quick enough? Although I'm not convinced it's how you are supposed to do things when you've explicltly configured multiple periods. I wonder what is the downside to this other than unnecessary interrupts? Potentially underunning?

Reply:
Good summary. Like you say, the size calculation I put in, while working alright, is in hindsight odd and suboptimal.

> I wonder what is the downside to this other than unnecessary interrupts? Potentially underunning?

Yes I think that a smaller internal buffer - and so more frequent "top-ups" - will lower the risk of underruns under load. Under load we cannot guarantee to fill the buffer quick enough, but at least we have more of a chance when we can write smaller chunks.

Reply:
> Sorry, I am a bit confused by this. We don't have a period-sized buffer in the old code or your new version. What seems to work is a buffer-sized buffer!

I just haven't changed it yet. I was waiting for @roderickvd's thoughts on it. A period sized buffer works just fine an was what was used before @roderickvd changed it. When I get off work I will change it to period sized.

Reply:
> A period sized buffer works just fine and was what was used before @roderickvd changed it.

Got you - the old-old code!

Reply:
I can do that also.

Reply:
> 
> 
> > I looked into it more a bit later for a contribution to RustAudio/cpal and it turns out that by default the value for the period size is the minimum period size. The effect: Basically a low wakeup interval with high latency.
> 
> The low wakeup interval makes sense. If you set a reasonable sized buffer with min period size that would make for _**a lot**_ of periods. I'm not sure why it would effect latency though? The total buffer size is the latency, so a 0.5 sec buffer with 4 periods or 10 periods is still just 0.5 sec of latency.

No it doesn't, but you don't really gain anything from the frequent delivery of data. You get  a high load with no benefits, since the latency is still half a second. That's what I meant.

At this moment alsa pulse still gives about 50ms (this varies a bit on the precise parameters involved. 33ms is also a common value) period time if I do not specify a period size for a 500ms buffer. That's 99 periods!

Reply:
The buffer is now just a period again.

Reply:
> No it doesn't, but you don't really gain anything from the frequent requests for data. You get a high load with no benefits, since the latency is still half a second. That's what I meant.

That was understanding, that it's a trade off between underruns and interrupts. 

At this moment alsa pulse still gives about 5ms (this varies a bit on the precise parameters involved) period time if I do not specify a period size/time for a 500ms buffer. That's 99 periods! And that 5ms is way to low for pulse with my pi B+.

NOTE: Fixed my math. Pulse hwparams even claims .333ms as its (non-inclusive) lower bound. 16 frames as its inclusive lower bound. That lower bound changes depending on the channels and the format.

I wonder if that's user configurable in PulseAudio? It is in ALSA using `dmix`:

 https://gist.github.com/JasonLG1979/f29cb0f14218a3b20db5d32c2cfae5ed
 
 You can manually configure every aspect of the buffer. Once configured that's what all apps get no matter what they ask for.